src/node_watchdog.cc

```diff
@@ -100,16 +124,55 @@ SigintWatchdog::SigintWatchdog(
   SigintWatchdogHelper::GetInstance()->Start();
 }
 
+SigintWatchdog::SigintWatchdog(
+  v8::Isolate* isolate, Local<v8::Function> interrupt)
+    : isolate_(isolate) {
+  ref.Reset(isolate, interrupt);
+  auto env = node::Environment::GetCurrent(isolate);
+  loop_activate = new uv_async_t;
+  loop_activate->data = this;
+  // Just activate the loop, doesn't do callback.
+  uv_async_init(env->event_loop(), loop_activate, [](uv_async_t* it){
+    SigintWatchdog* self = reinterpret_cast<SigintWatchdog*>(it->data);
+    self->HandleInterrupt();
+  });
+  uv_unref(reinterpret_cast<uv_handle_t*>(loop_activate));
+  // Register this watchdog with the global SIGINT/Ctrl+C listener.
+  SigintWatchdogHelper::GetInstance()->Register(this);
+  // Start the helper thread, if that has not already happened.
+  SigintWatchdogHelper::GetInstance()->Start();
+}
 
 SigintWatchdog::~SigintWatchdog() {
+  if (loop_activate != nullptr) {
+    uv_close(reinterpret_cast<uv_handle_t*>(loop_activate),
+      SigintWatchdog::CloseAsync);
+  }
+  ref.Reset();
   SigintWatchdogHelper::GetInstance()->Unregister(this);
   SigintWatchdogHelper::GetInstance()->Stop();
 }
 
-
 void SigintWatchdog::HandleSigint() {
-  *received_signal_ = true;
-  isolate_->TerminateExecution();
+  if (ref.IsEmpty()) {
+    *received_signal_ = true;
+    isolate_->TerminateExecution();
+  } else {
+    uv_async_send(loop_activate);
+    isolate_->RequestInterrupt([](v8::Isolate* isolate, void* data) {
```

Comment:
I don’t think it’s technically legal to do this (call JS from an interrupt handler), but we also already do it for the inspector? I’d at least check that the two don’t get in each others way, e.g. that you can set a breakpoint inside the JS handler or something…

Reply:
I believe it's allowed. The relevant code is here:

https://github.com/nodejs/node/blob/edc83a9203c2a8d0fd219aac95fd1615e001c484/deps/v8/src/execution/isolate.cc#L1429-L1445

Note the explicit switch to `VMState<EXTERNAL>`.